be/queue.c

```diff
@@ -0,0 +1,443 @@
+		m0_free(bq->bq_ops_get);
+		m0_free(bq->bq_ops_put);
+		m0_free(bq->bq_qitems);
+		return M0_ERR(-ENOMEM);
+	}
+	m0_mutex_init(&bq->bq_lock);
+	bqop_tlist_init(&bq->bq_op_put_unused);
+	for (i = 0; i < bq->bq_cfg.bqc_producers_nr_max; ++i) {
+		bqop_tlink_init_at_tail(&bq->bq_ops_put[i],
+					 &bq->bq_op_put_unused);
+	}
+	bqop_tlist_init(&bq->bq_op_put);
+	bqop_tlist_init(&bq->bq_op_get_unused);
+	for (i = 0; i < bq->bq_cfg.bqc_consumers_nr_max; ++i) {
+		bqop_tlink_init_at_tail(&bq->bq_ops_get[i],
+					 &bq->bq_op_get_unused);
+	}
+	bqop_tlist_init(&bq->bq_op_get);
+	bqq_tlist_init(&bq->bq_q_unused);
+	for (i = 0; i < be_queue_qitems_nr(bq); ++i) {
+		bqq_tlink_init_at_tail(be_queue_qitem(bq, i),
+				       &bq->bq_q_unused);
+	}
+	bqq_tlist_init(&bq->bq_q);
+	return M0_RC(0);
+}
+
+M0_INTERNAL void m0_be_queue_fini(struct m0_be_queue *bq)
+{
+	struct be_queue_wait_op  *bwo;
+	struct be_queue_item     *bqi;
+	uint64_t                i;
+
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	M0_ASSERT_INFO(bq->bq_enqueued == bq->bq_dequeued,
+	               "bq="BEQ_F, BEQ_P(bq));
+
+	m0_tl_for(bqq, &bq->bq_q, bqi) {
+		M0_LOG(M0_ERROR, "there is an item in the queue"); /* XXX */
+	} m0_tl_endfor;
+	bqq_tlist_fini(&bq->bq_q);
+	m0_tl_for(bqop, &bq->bq_op_get, bwo) {
+		M0_LOG(M0_ERROR, "bq=%p bbo_data="BUF_F,
+		       bq, BUF_P(&bwo->bbo_data));
+	} m0_tl_endfor;
+	bqop_tlist_fini(&bq->bq_op_get);
+	for (i = 0; i < bq->bq_cfg.bqc_consumers_nr_max; ++i)
+		bqop_tlink_del_fini(&bq->bq_ops_get[i]);
+	bqop_tlist_fini(&bq->bq_op_get_unused);
+	/* if there was nothing in bq_q then the following list is empty */
+	bqop_tlist_fini(&bq->bq_op_put);
+	for (i = 0; i < bq->bq_cfg.bqc_producers_nr_max; ++i)
+		bqop_tlink_del_fini(&bq->bq_ops_put[i]);
+	bqop_tlist_fini(&bq->bq_op_put_unused);
+	for (i = 0; i < be_queue_qitems_nr(bq); ++i)
+		bqq_tlink_del_fini(be_queue_qitem(bq, i));
+	bqq_tlist_fini(&bq->bq_q_unused);
+	m0_mutex_fini(&bq->bq_lock);
+	m0_free(bq->bq_ops_put);
+	m0_free(bq->bq_ops_get);
+	m0_free(bq->bq_qitems);
+	M0_LEAVE();
+}
+
+M0_INTERNAL void m0_be_queue_lock(struct m0_be_queue *bq)
+{
+	m0_mutex_lock(&bq->bq_lock);
+}
+
+M0_INTERNAL void m0_be_queue_unlock(struct m0_be_queue *bq)
+{
+	m0_mutex_unlock(&bq->bq_lock);
+}
+
+static uint64_t be_queue_q_size(struct m0_be_queue *bq)
+{
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_ASSERT_INFO(bq->bq_enqueued >= bq->bq_dequeued,
+	               "bq="BEQ_F, BEQ_P(bq));
+	return bq->bq_enqueued - bq->bq_dequeued;
+}
+
+static bool be_queue_is_empty(struct m0_be_queue *bq)
+{
+	return be_queue_q_size(bq) == 0;
+}
+
+static bool be_queue_is_full(struct m0_be_queue *bq)
+{
+	return be_queue_q_size(bq) >= bq->bq_cfg.bqc_q_size_max;
+}
+
+static struct be_queue_item *be_queue_q_put(struct m0_be_queue  *bq,
+                                            const struct m0_buf *data)
+{
+	struct be_queue_item *bqi;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+
+	bqi = bqq_tlist_head(&bq->bq_q_unused);
```

Comment:
What if this returns NULL?
That means there is no unused item.

Reply:
Do we need to remove this `bqi` from the `bq_q_unused` list?

Reply:
> That means there is no unused item.

This would be a failure of the user to provide a correct configuration (# of items in the queue + # of producers), so the code would crash.

Reply:
> The bqq_tlist_move_tail will remove the bqi from old list and move to new list.

Right.

Reply:
Then, I propose to add:
M0_ASSERT_INFO(bqi != NULL, "Too many producers");

Reply:
Done.